pytest_fxa/plugin.py

```diff
@@ -74,20 +89,25 @@ def fxa_account(fxa_client, fxa_email):
     password = ''.join([random.choice(string.ascii_letters) for i in range(8)])
     FxAccount = collections.namedtuple('FxAccount', 'email password')
     fxa_account = FxAccount(email=account.email, password=password)
-    session = fxa_client.create_account(fxa_account.email,
-                                        fxa_account.password)
-    logger.info('Created: {}'.format(fxa_account))
-    account.fetch()
-    message = account.wait_for_email(lambda m: 'x-verify-code' in m['headers'])
-    session.verify_email_code(message['headers']['x-verify-code'])
-    logger.info('Verified: {}'.format(fxa_account))
-    yield fxa_account
     try:
-        account.clear()
-        fxa_client.destroy_account(fxa_account.email, fxa_account.password)
-        logger.info('Removed: {}'.format(fxa_account))
-    except ClientError as e:
-        # 'Unknown Account' error is ok - account already deleted
-        # https://github.com/mozilla/fxa-auth-server/blob/master/docs/api.md#response-format
-        if e.errno not in [102]:
-            raise
+        session = fxa_client.create_account(fxa_account.email,
+                                            fxa_account.password)
+        logger.info('Created: {}'.format(fxa_account))
+        account.fetch()
+        message = account.wait_for_email(
+            lambda m: 'x-verify-code' in m['headers'] and
+            session.uid == m['headers']['x-uid']
+        )
+        session.verify_email_code(message['headers']['x-verify-code'])
+        logger.info('Verified: {}'.format(fxa_account))
+    except ClientError:
+        raise
+    except TypeError:
+        # Account didn't get verified, server didn't respond or returned `None`
+        raise RuntimeError("The account {} could not be verified.".format(
+            fxa_account.email)
+        )
```

Comment:
This does improve the error message, but I'm concerned it may be misleading. I think what you'd want to do is directly check if `message` is `None` and log something appropriate or raise an exception. Preferably, PyFxA would raise a suitable exception when calling `wait_for_email` instead of returning `None`.

Reply:
I added this mainly as a Timeout. PyFxA doesn't raise anything if the verification returns back as ```None```. If ```wait_for_email``` times out it will return a ```None```. See [this](https://github.com/mozilla/PyFxA/blob/b2fda137eeaf1a3e428a27a68508df5bd9fcb2f7/fxa/tests/utils.py#L99). I experienced this testing on addons-server.

Reply:
Something like this? 
```
except TypeError:
        # Account didn't get verified, server didn't respond or returned `None`
        if account.wait_for_email is None:
            raise RuntimeError("The account {} could not be verified.".format(
                fxa_account.email)
            )
        else:
            raise
```

Reply:
No need to catch the exception if you're just going to raise it again. You can just check if `message` is `None` and raise a custom exception (preferred to `RuntimeError`) or log an error, which will show up in the failure details.

Reply:
```Message``` won't be none at that point. It will have the response with something like this:
```
[{'html': '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">\n<html xmlns="http://www.w3.org/1999/xhtml">\n<head>\n  <meta http-equiv="content-type" content="text/html; charset=utf-8" />\n  <title>Firefox Accounts</title>\n</head>\n\n<body style="-ms-text-size-adjust: 100%; -webkit-text-size-adjust: 100%; margin: 0; padding: 0;">\n<table align="center" border="0" cellpadding="0" cellspacing="0" width="310" style="-webkit-text-size-adjust: 100%; border-collapse: collapse; mso-table-lspace: 0pt; mso-table-rspace: 0pt; width: 310px; margin: 0 auto;">\n\n<!--Logo-->\n<tr style="page-break-before: always">\n  <td align="center" id="firefox-logo" style="padding: 20px 0;">\n      <img src="https://image.e.mozilla.org/lib/fe9915707361037e75/m/3/firefox57-logo.png" height="88" width="85" alt="" style="-ms-interpolation-mode: bicubic;" />\n\n  </td>\n</tr>\n\n\n<!--Header Area-->\n<tr style="page-break-before: always">\n  <td valign="top">\n    <h1 style="font-family: sans-serif; font-size: 21px; line-height: 29px; font-weight: normal; margin: 0 0 11px 0; text-align: center;">Welcome!</h1>\n    <p class="primary" style="font-family: sans-serif; font-size: 14px; line-height: 21px; font-weight: normal; margin: 0 0 21px 0; text-align: center;">Please activate your account by confirming this email address.</p>\n  </td>\n</tr>\n\n<!--Button Area-->\n<tr height="50">\n  <td align="center" valign="top">\n    <table border="0" cellpadding="0" cellspacing="0" height="100%" width="100%" id="email-button" style="-webkit-text-size-adjust: 100%; border-collapse: collapse; mso-table-lspace: 0pt; mso-table-rspace: 0pt; background-color: #0a84ff; border-radius: 4px; height: 50px; width: 310px !important;">\n      <tr style="page-break-before: always">\n        <td align="center" valign="middle" id="button-content" style="font-family: sans-serif; font-weight: normal; text-align: center; margin: 0; color: #ffffff; font-size: 20px; line-height: 100%;">\n          <!--[if mso]>\n          <v:roundrect xmlns:v="urn:schemas-microsoft-com:vml" xmlns:w="urn:schemas-microsoft-com:office:word" href="https://accounts.stage.mozaws.net/verify_email?uid=bdeae34be5d646bd9ef4eab1b9b39c13&code=59bab55339c4c90d0b4f3e95880cb3d4&utm_medium=email&utm_campaign=fx-welcome&utm_content=fx-activate" style="width:280px;height:40px;v-text-anchor:middle;" arcsize="10%" stroke="f" fillcolor="#0a84ff">\n            <w:anchorlock/>\n            <center>\n          <![endif]-->\n          <a href="https://accounts.stage.mozaws.net/verify_email?uid=bdeae34be5d646bd9ef4eab1b9b39c13&code=59bab55339c4c90d0b4f3e95880cb3d4&utm_medium=email&utm_campaign=fx-welcome&utm_content=fx-activate" id="button-link" style="font-family:sans-serif; color: #fff; display: block; padding: 15px; text-decoration: none; width: 280px; font-size: 18px; line-height: 26px;">Activate now</a>\n          <!--[if mso]>\n          </center>\n          </v:roundrect>\n          <![endif]-->\n        </td>\n      </tr>\n    </table>\n  </td>\n</tr>\n<!--Button Area-->\n<tr style="page-break-before: always">\n  <td border="0" cellpadding="0" cellspacing="0" height="100%" width="100%">\n    <br/>\n    <p class="secondary" style="font-family: sans-serif; font-weight: normal; margin: 0 0 12px 0; text-align: center; color: #737373; font-size: 11px; line-height: 18px; width: 310px !important; word-wrap:break-word">This is an automated email; if you received it in error, no action is required. For more information, please visit <a href="https://support.mozilla.org/kb/im-having-problems-with-my-firefox-account?utm_medium=email&utm_campaign=fx-welcome&utm_content=fx-support" style="color: #0a84ff; text-decoration: none; font-family: sans-serif;">Mozilla Support</a>.</p>\n  </td>\n</tr>\n\n\n<tr style="page-break-before: always">\n  <td valign="top">\n    <p style="font-family: sans-serif; font-weight: normal; margin: 0; text-align: center; color: #737373; font-size: 11px; line-height: 18px; width: 310px !important; word-wrap:break-word">Mozilla. 331 E Evelyn Ave, Mountain View, CA 94041\n    <br />\n    <a href="https://www.mozilla.org/privacy?utm_medium=email&utm_campaign=fx-welcome&utm_content=fx-privacy" style="color: #0a84ff; text-decoration: none; font-family: sans-serif; font-size: 11px; line-height: 18px;">Mozilla Privacy Policy</a></p>\n  </td>\n</tr>\n\n</table>\n\n<div itemscope itemtype="https://schema.org/EmailMessage">\n  <div itemprop="potentialAction" itemscope itemtype="https://schema.org/ViewAction">\n    <link itemprop="target" href="https://accounts.stage.mozaws.net/verify_email?uid=bdeae34be5d646bd9ef4eab1b9b39c13&code=59bab55339c4c90d0b4f3e95880cb3d4&one_click=true&utm_medium=email&utm_campaign=fx-welcome&utm_content=fx-activate-oneclick"/>\n    <meta itemprop="name" content="Verify Email"/>\n    <meta itemprop="url" content="https://accounts.stage.mozaws.net/verify_email?uid=bdeae34be5d646bd9ef4eab1b9b39c13&code=59bab55339c4c90d0b4f3e95880cb3d4&one_click=true&utm_medium=email&utm_campaign=fx-welcome&utm_content=fx-activate-oneclick"/>\n  </div>\n  <meta itemprop="description" content="Verify your email to finish your Firefox Account registration"/>\n</div>\n\n</body>\n</html>\n', 'text': 'Welcome!\n\nPlease activate your account by confirming this email address.\nActivate now: https://accounts.stage.mozaws.net/verify_email?uid=bdeae34be5d646bd9ef4eab1b9b39c13&code=59bab55339c4c90d0b4f3e95880cb3d4&utm_medium=email&utm_campaign=fx-welcome&utm_content=fx-activate\n\nThis is an automated email; if you received it in error, no action is required. For more information, please visit https://support.mozilla.org/kb/im-having-problems-with-my-firefox-account?utm_medium=email&utm_campaign=fx-welcome&utm_content=fx-support\n\nMozilla. 331 E Evelyn Ave, Mountain View, CA 94041\nMozilla Privacy Policy https://www.mozilla.org/privacy?utm_medium=email&utm_campaign=fx-welcome&utm_content=fx-privacy\n', 'headers': {'dkim-signature': 'v=1; a=rsa-sha256; q=dns/txt; c=relaxed/simple; s=ug7nbtf4gccmlpwj322ax3p6ow6yfsug; d=amazonses.com; t=1534181022; h=Content-Type:From:Sender:To:Subject:Message-ID:Date:MIME-Version:Feedback-ID; bh=6OGrvhLJKqOgGHO9Y48vMuyiyQss7eQvl+ShzJJZ48Q=; b=EmfHbhPAqAbXUvxwG7WRZUhJWMqXHuA9nUQn8Mj/b/sDbRH12iGCP9+VflzvyBLp cmyN13rLOpQiKI19Wt9SIvsM4M8A+eCVzzOwa11b1Kz75xXPzM1uUVq3Uc4nhkPUwxI /gSd9AdofQ2HH83ld2+fnwZxUSVbfoWokl2ECDQI=', 'content-type': 'multipart/alternative; boundary="----sinikael-?=_1-15341810218370.739816807310687"', 'content-language': 'en', 'x-template-name': 'verifyEmail', 'x-template-version': '1', 'x-link': 'https://accounts.stage.mozaws.net/verify_email?uid=bdeae34be5d646bd9ef4eab1b9b39c13&code=59bab55339c4c90d0b4f3e95880cb3d4&utm_medium=email&utm_campaign=fx-welcome&utm_content=fx-activate', 'x-verify-code': '59bab55339c4c90d0b4f3e95880cb3d4', 'x-uid': 'bdeae34be5d646bd9ef4eab1b9b39c13', 'x-email-service': 'fxa-auth-server', 'x-email-sender': 'ses', 'from': 'Firefox Accounts <verification@stage.mozaws.net>', 'sender': 'Firefox Accounts <verification@stage.mozaws.net>', 'to': 'pytest-5b5092fdc1@restmail.net', 'subject': 'Verify your Firefox Account', 'message-id': '<010001653450990e-9d2fa7a2-b0c5-479e-9fb4-a57e4704ba4e-000000@email.amazonses.com>', 'date': 'Mon, 13 Aug 2018 17:23:42 +0000', 'mime-version': '1.0', 'x-ses-outgoing': '2018.08.13-54.240.9.110', 'feedback-id': '1.us-east-1.hwss4XO2uXvrnVmJ1k8X4mRPkexq3Zj6phfvp+oBp9Q=:AmazonSES'}, 'subject': 'Verify your Firefox Account', 'messageId': '010001653450990e-9d2fa7a2-b0c5-479e-9fb4-a57e4704ba4e-000000@email.amazonses.com', 'priority': 'normal', 'from': [{'address': 'verification@stage.mozaws.net', 'name': 'Firefox Accounts'}], 'to': [{'address': 'pytest-5b5092fdc1@restmail.net', 'name': ''}], 'date': '2018-08-13T17:23:42.000Z', 'receivedAt': '2018-08-13T17:23:42.274Z'}]
```

Reply:
We're going back and forward on something quite straight-forward 😄 The `TypeError` is only raised because you're trying to do something with `None` that isn't supported. Therefore you can simply do `if message is None` and within that block log an error that helps to identify the cause of the test failure, or raise a custom exception.

Reply:
If I except ```TypeError``` and it isn't ```None``` shouldn't I raise that? Isn't that what you said I should do just incase there is another ```TypeError``` somewhere else?

Reply:
No, I have been suggesting that you remove the `except TypeError` entirely and simply checking if the value of `message` is `None`. That way any exceptions that occur will be raised as we would expect. Sorry if this hasn't been clear.

Reply:
Ah well this is software. I think back and forth is kinda going to happen 😁. Thanks for clearing it up.